### PR TITLE
[Fix] Broken Build Scripts

### PR DESCRIPTION
**Pull request by chesspro13**
Problem:
Build scripts were broken with migration to Typescript.

Solution:
Use tsc to compile TS back into JS for execution.

Tested and working:

- [x] build-debian
- [x] build-docker
- [x] build-linux64
- [ ] build-mac-x64
- [ ] build-win-x64

### CODE REVIEW

**Comment by rauenzi** [comment]
I get that it's a minor thing (and private to the build process) but I think `compiledJS` is probably not the best name for the folder. If `build` isn't already used, that would be good.

**Comment by perfectra1n** [comment]
```sh
(v22.4.0) [root on DESKTOP-M0FBO5A] ~/repos/TriliumNext/Notes on test/do:test/do
»  docker build .
[+] Building 29.1s (12/12) FINISHED                                                                                                                                                                                                                                      docker:default
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                               0.0s
 => => transferring dockerfile: 1.21kB                                                                                                                                                                                                                                             0.0s
 => [internal] load metadata for docker.io/library/node:18.18.2-alpine                                                                                                                                                                                                             0.1s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                  0.0s
 => => transferring context: 94B                                                                                                                                                                                                                                                   0.0s
 => [1/7] FROM docker.io/library/node:18.18.2-alpine@sha256:16b46e5ea9fb5c2d13dda36f0feb670fa89de6a412725007555f2eee9a126b60                                                                                                                                                       1.8s
 => => resolve docker.io/library/node:18.18.2-alpine@sha256:16b46e5ea9fb5c2d13dda36f0feb670fa89de6a412725007555f2eee9a126b60                                                                                                                                                       0.0s
 => => sha256:bd5b864d12491e8d318149d93e4c17adc97089ca47f3747b2fad280f62002181 451B / 451B                                                                                                                                                                                         0.2s
 => => sha256:16b46e5ea9fb5c2d13dda36f0feb670fa89de6a412725007555f2eee9a126b60 1.43kB / 1.43kB                                                                                                                                                                                     0.0s
 => => sha256:a2c1bfbc28ede8a10b48fd6889446e04c5c362219fa785633a452d77a1fdd5f1 1.16kB / 1.16kB                                                                                                                                                                                     0.0s
 => => sha256:1646380c31566e4d7f020d812298baab2b1b67272ba0281b4f37141393b73f65 7.14kB / 7.14kB                                                                                                                                                                                     0.0s
 => => sha256:3ebae63ae1d8896e6d3d75cb0ee2aa5fc33676028241c1130d1729fe974b4a2f 39.97MB / 39.97MB                                                                                                                                                                                   0.7s
 => => sha256:4df6e7e40a3fc8572a44be8b3ed0cd78c2a5fc6968441f4ac7164d240919d9cd 2.34MB / 2.34MB                                                                                                                                                                                     0.2s
 => => extracting sha256:3ebae63ae1d8896e6d3d75cb0ee2aa5fc33676028241c1130d1729fe974b4a2f                                                                                                                                                                                          0.8s
 => => extracting sha256:4df6e7e40a3fc8572a44be8b3ed0cd78c2a5fc6968441f4ac7164d240919d9cd                                                                                                                                                                                          0.0s
 => => extracting sha256:bd5b864d12491e8d318149d93e4c17adc97089ca47f3747b2fad280f62002181                                                                                                                                                                                          0.0s
 => [internal] load build context                                                                                                                                                                                                                                                  0.0s
 => => transferring context: 64.54kB                                                                                                                                                                                                                                               0.0s
 => [2/7] WORKDIR /usr/src/app                                                                                                                                                                                                                                                     0.2s
 => [3/7] COPY . .                                                                                                                                                                                                                                                                 0.2s
 => [4/7] COPY server-package.json package.json                                                                                                                                                                                                                                    0.1s
 => [5/7] RUN set -x     && apk add --no-cache --virtual .build-dependencies         autoconf         automake         g++         gcc         libtool         make         nasm         libpng-dev         python3     && npm install     && apk del .build-dependencies     &&  23.9s
 => [6/7] RUN apk add --no-cache su-exec shadow                                                                                                                                                                                                                                    0.5s
 => [7/7] RUN adduser -s /bin/false node; exit 0                                                                                                                                                                                                                                   0.4s
 => exporting to image                                                                                                                                                                                                                                                             1.7s
 => => exporting layers                                                                                                                                                                                                                                                            1.7s
 => => writing image sha256:28378617ab1cb20643067dee78b79db17140f41bed5f044b89dcf37a01b20191 
```

It's a beautiful thing!

**Comment by eliandoran** [comment]
@chesspro13 , on my side the docker image appears to build fine, but when running it with no data directory it fails with:

```
Error: Failed to lookup view "setup" in views directory "/usr/src/app/src/views"
    at Function.render (/usr/src/app/node_modules/express/lib/application.js:597:17)
    at ServerResponse.render (/usr/src/app/node_modules/express/lib/response.js:1039:7)
    at setupPage (/usr/src/app/src/routes/setup.js:27:9)
    at Function.cb (/usr/src/app/src/routes/routes.js:400:51)
    at Function.sqliteTransaction (/usr/src/app/node_modules/better-sqlite3/lib/methods/transaction.js:65:24)
    at Object.transactional (/usr/src/app/src/services/sql.js:274:50)
    at /usr/src/app/src/routes/routes.js:401:48
    at /usr/src/app/node_modules/cls-hooked/context.js:118:13
    at Namespace.run (/usr/src/app/node_modules/cls-hooked/context.js:97:5)
    at Namespace.runAndReturn (/usr/src/app/node_modules/cls-hooked/context.js:117:8) {
  view: View {
    defaultEngine: 'ejs',
    ext: '.ejs',
    name: 'setup',
    root: '/usr/src/app/src/views',
    engine: [Function (anonymous)],
    path: undefined
  }
}```

I'll see if I can track down the issue.

**Comment by chesspro13** [comment]
> @chesspro13 , on my side the docker image appears to build fine, but when running it with no data directory it fails with:
> 
> ```
> Error: Failed to lookup view "setup" in views directory "/usr/src/app/src/views"
>     at Function.render (/usr/src/app/node_modules/express/lib/application.js:597:17)
>     at ServerResponse.render (/usr/src/app/node_modules/express/lib/response.js:1039:7)
>     at setupPage (/usr/src/app/src/routes/setup.js:27:9)
>     at Function.cb (/usr/src/app/src/routes/routes.js:400:51)
>     at Function.sqliteTransaction (/usr/src/app/node_modules/better-sqlite3/lib/methods/transaction.js:65:24)
>     at Object.transactional (/usr/src/app/src/services/sql.js:274:50)
>     at /usr/src/app/src/routes/routes.js:401:48
>     at /usr/src/app/node_modules/cls-hooked/context.js:118:13
>     at Namespace.run (/usr/src/app/node_modules/cls-hooked/context.js:97:5)
>     at Namespace.runAndReturn (/usr/src/app/node_modules/cls-hooked/context.js:117:8) {
>   view: View {
>     defaultEngine: 'ejs',
>     ext: '.ejs',
>     name: 'setup',
>     root: '/usr/src/app/src/views',
>     engine: [Function (anonymous)],
>     path: undefined
>   }
> }
> ```
> 
> I'll see if I can track down the issue.

TSC doesn't move the .ejs files so the script is supposed to move those over manually. Let me double check. I fixed it, but may have reverted it.

**Comment by eliandoran** [comment]
> > @chesspro13 , on my side the docker image appears to build fine, but when running it with no data directory it fails with:
> > ```
> > Error: Failed to lookup view "setup" in views directory "/usr/src/app/src/views"
> >     at Function.render (/usr/src/app/node_modules/express/lib/application.js:597:17)
> >     at ServerResponse.render (/usr/src/app/node_modules/express/lib/response.js:1039:7)
> >     at setupPage (/usr/src/app/src/routes/setup.js:27:9)
> >     at Function.cb (/usr/src/app/src/routes/routes.js:400:51)
> >     at Function.sqliteTransaction (/usr/src/app/node_modules/better-sqlite3/lib/methods/transaction.js:65:24)
> >     at Object.transactional (/usr/src/app/src/services/sql.js:274:50)
> >     at /usr/src/app/src/routes/routes.js:401:48
> >     at /usr/src/app/node_modules/cls-hooked/context.js:118:13
> >     at Namespace.run (/usr/src/app/node_modules/cls-hooked/context.js:97:5)
> >     at Namespace.runAndReturn (/usr/src/app/node_modules/cls-hooked/context.js:117:8) {
> >   view: View {
> >     defaultEngine: 'ejs',
> >     ext: '.ejs',
> >     name: 'setup',
> >     root: '/usr/src/app/src/views',
> >     engine: [Function (anonymous)],
> >     path: undefined
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'll see if I can track down the issue.
> 
> TSC doesn't move the .ejs files so the script is supposed to move those over manually. Let me double check. I fixed it, but may have reverted it.

@chesspro13 , I think it's a mistake on my side. I ran `bin/build-docker.sh` directly without calling `bin/build.sh` (which seems to do `./copy-trilium.sh`). Still waiting for the build to finish to tell if it's actually working.

**Comment by chesspro13** [comment]
> > > @chesspro13 , on my side the docker image appears to build fine, but when running it with no data directory it fails with:
> > > ```
> > > Error: Failed to lookup view "setup" in views directory "/usr/src/app/src/views"
> > >     at Function.render (/usr/src/app/node_modules/express/lib/application.js:597:17)
> > >     at ServerResponse.render (/usr/src/app/node_modules/express/lib/response.js:1039:7)
> > >     at setupPage (/usr/src/app/src/routes/setup.js:27:9)
> > >     at Function.cb (/usr/src/app/src/routes/routes.js:400:51)
> > >     at Function.sqliteTransaction (/usr/src/app/node_modules/better-sqlite3/lib/methods/transaction.js:65:24)
> > >     at Object.transactional (/usr/src/app/src/services/sql.js:274:50)
> > >     at /usr/src/app/src/routes/routes.js:401:48
> > >     at /usr/src/app/node_modules/cls-hooked/context.js:118:13
> > >     at Namespace.run (/usr/src/app/node_modules/cls-hooked/context.js:97:5)
> > >     at Namespace.runAndReturn (/usr/src/app/node_modules/cls-hooked/context.js:117:8) {
> > >   view: View {
> > >     defaultEngine: 'ejs',
> > >     ext: '.ejs',
> > >     name: 'setup',
> > >     root: '/usr/src/app/src/views',
> > >     engine: [Function (anonymous)],
> > >     path: undefined
> > >   }
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I'll see if I can track down the issue.
> > 
> > 
> > TSC doesn't move the .ejs files so the script is supposed to move those over manually. Let me double check. I fixed it, but may have reverted it.
> 
> @chesspro13 , I think it's a mistake on my side. I ran `bin/build-docker.sh` directly without calling `bin/build.sh` (which seems to do `./copy-trilium.sh`). Still waiting for the build to finish to tell if it's actually working.

@eliandoran `bin/build.sh` doesn't build for docker. `bin/build-docker.sh` is correct. Dockerfile does `bin/copy-trilium.sh`'s job.

**Comment by eliandoran** [comment]
@chesspro13 , you're right. It does seem that our Docker build does not have the `/usr/src/app/src/views` directory. I've just tested on zadam/trilium on clean and the folder is there without issue. Not sure exactly why we don't have it.

**Comment by chesspro13** [comment]
> on zadam/trilium on clean and the folder is there without issue. Not sure exactly why we don't hav

@eliandoran I just found that too. I guess I never updated the Dockerfile. Doing that now.

File: bin/build-docker.sh [hunk 1]
```diff
@@ -3,7 +3,14 @@
 VERSION=`jq -r ".version" package.json`
 SERIES=${VERSION:0:4}-latest
 
-cat package.json | grep -v electron > server-package.json
+if ! [[ $(which tsc) ]]; then
+    npm i -g typescript
+fi
+
+cat package.json > server-package.json
```
Comment (eliandoran) [inline bin/build-docker.sh:10]: Why was `grep -v electron` removed?

I suppose this is helping to reduce the size of the Docker image by not embedding electron dependencies.

File: bin/copy-trilium.sh [hunk 2]
```diff
@@ -18,16 +24,20 @@ mkdir -pv "$DIR"
...
-for f in 'package.json' 'package-lock.json' 'README.md' 'LICENSE' 'config-sample.ini' 'electron.js'; do
+for f in 'README.md' 'LICENSE' 'config-sample.ini'; do
     cp "$f" "$DIR"/
 done
-cp webpack-* "$DIR"/      # here warning because there is no 'webpack-*', but webpack.config.js only
+cp webpack-* "$DIR"/      # here warning because there is no 'webpack-*', but webpack.config.ts only
 
 # run in subshell (so we return to original dir)
-(cd $DIR && n exec 18.18.2 npm install --only=prod)
+# (cd $DIR && n exec 18.18.2 npm install --omit=dev && n exec 18.18.2 npm rebuild)
```
Comment (eliandoran) [inline bin/copy-trilium.sh:39]: Please remove the commented-out code. Personally I think it's OK to remove the dependency to `n` for managing the npm version.

File: bin/build-docker.sh [hunk 3]
```diff
@@ -3,7 +3,14 @@
 VERSION=`jq -r ".version" package.json`
 SERIES=${VERSION:0:4}-latest
 
-cat package.json | grep -v electron > server-package.json
+if ! [[ $(which tsc) ]]; then
+    npm i -g typescript
+fi
```
Comment (eliandoran) [inline bin/build-docker.sh:8]:
```suggestion
```

Not needed, since we can use `npx`.

File: bin/copy-trilium.sh [hunk 4]
```diff
@@ -9,6 +9,12 @@ if ! [[ $(which npm) ]]; then
     exit 1
 fi
 
+if ! [[ $(which tsc) ]]; then
+    npm i -g typescript
+fi
+
+tsc
```
Comment (eliandoran) [inline bin/copy-trilium.sh:16]:
```suggestion
npx tsc
```

File: bin/build-docker.sh [hunk 5]
```diff
@@ -3,7 +3,14 @@
...
+
+cat package.json > server-package.json
+
+echo "Compiling typescript..."
+tsc
```
Comment (eliandoran) [inline bin/build-docker.sh:13]:
```suggestion
npx tsc
```

File: libraries/ckeditor/ckeditor.js.map
Comment (eliandoran) [inline libraries/ckeditor/ckeditor.js.map:1]: Please revert changes to this file.

File: package-lock.json
Comment (eliandoran) [inline package-lock.json:1]: Please revert whitespace changes to this file.

File: package.json
Comment (eliandoran) [inline package.json:1]: Please revert whitespace changes to this file.

File: tsconfig.json
Comment (eliandoran) [inline tsconfig.json:1]: Please revert whitespace changes to this file.

**Review by eliandoran — CHANGES_REQUESTED** [review]

**Comment by chesspro13** [comment]
@eliandoran looks like there were still a few files that were needed for the docker build. Updated the `Dockerfile` and it builds and runs. Gonna work on the rest of the requests!

File: bin/build-docker.sh (hunk 1 above)
Comment (chesspro13) [reply bin/build-docker.sh:10]: I added it back. It was causing problems by removing the `postinstall` script causing errors in testing. Later found out @alexpietsch added it a while ago, but it doesn't seam to be need. I ended up just getting rid of the reference to electron in the `postinstall` script.

File: bin/copy-trilium.sh (hunk 2 above)
Comment (chesspro13) [reply bin/copy-trilium.sh:39]: Removed. Nice catch!

File: bin/build-docker.sh (hunk 3 above)
Comment (chesspro13) [reply bin/build-docker.sh:8]: This was added to make sure that tsc could run and convert all the `.ts` to `.js` which removed all the "CommonJS" vs  "ESNext" errors I was getting. After making the rest of the changes, I'll come back to this and do some testing with `npx`.

Comment (chesspro13) [reply bin/build-docker.sh:8]: Removed.

File: bin/copy-trilium.sh (hunk 4 above)
Comment (chesspro13) [reply bin/copy-trilium.sh:16]: Done.

File: libraries/ckeditor/ckeditor.js.map
Comment (chesspro13) [reply libraries/ckeditor/ckeditor.js.map:1]: Only real change there was `webpack.config.js` -> `webpack.config.ts`.

**Comment by chesspro13** [comment]
Requested changes made, reopening PR.

File: webpack.config.ts
Comment (eliandoran) [inline webpack.config.ts:1]: I think changing this file is not necessary.

When developing #196 I had a similar issue:

```
[webpack-cli] Failed to load '/usr/src/app/webpack.config.js' config
[webpack-cli] Error: Cannot find module './src/services/asset_path'
```

This was actually a side effect of the Docker build copying over the .ts files instead of the built ones. 

I fixed it here:
https://github.com/TriliumNext/Notes/pull/196/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R24-R27

My recommendation would be to revert this change for now.

File: libraries/ckeditor/ckeditor.js.map
Comment (eliandoran) [inline libraries/ckeditor/ckeditor.js.map:1]: We should revert this change as well, since we don't need to modify the `ckeditor` (at least for now).

File: Dockerfile
Comment (eliandoran) [inline Dockerfile:1]: We can revert these changes as well, since I handle them in separately in https://github.com/TriliumNext/Notes/pull/196/ .

In your original implementation the `src` folder is deleted and reconstructed. However there are quite a few files to be copied, such as the webpack client artifacts that were not copied and would fail when the Docker image is run.

**Review by eliandoran — CHANGES_REQUESTED** [review]

**Comment by eliandoran** [comment]
@chesspro13 , this PR should be superseded by #196 and #201 .
Thank you for your contribution, since it really helped me kick start the development.
